tpm2/keydata_v3.go

```diff
@@ -0,0 +1,136 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019-2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package tpm2
+
+import (
+	"bytes"
+	"errors"
+	"io"
+
+	"github.com/canonical/go-tpm2"
+	"github.com/canonical/go-tpm2/mu"
+	"github.com/canonical/go-tpm2/util"
+
+	"golang.org/x/xerrors"
+)
+
+// keyData_v3 represents version 3 of keyData.
+type keyData_v3 struct {
+	KeyPrivate       tpm2.Private
+	KeyPublic        *tpm2.Public
+	KeyImportSymSeed tpm2.EncryptedSecret
+	PolicyData       *keyDataPolicy_v3
+}
+
+func readKeyDataV3(r io.Reader) (keyData, error) {
+	var d *keyData_v3
+	if _, err := mu.UnmarshalFromReader(r, &d); err != nil {
+		return nil, err
+	}
+	return d, nil
+}
+
+func (d *keyData_v3) Version() uint32 {
+	return 3
+}
+
+func (d *keyData_v3) Private() tpm2.Private {
+	return d.KeyPrivate
+}
+
+func (d *keyData_v3) Public() *tpm2.Public {
+	return d.KeyPublic
+}
+
+func (d *keyData_v3) ImportSymSeed() tpm2.EncryptedSecret {
+	return d.KeyImportSymSeed
+}
+
+func (d *keyData_v3) Imported(priv tpm2.Private) {
+	if d.KeyImportSymSeed == nil {
+		panic("does not need to be imported")
+	}
+	d.KeyPrivate = priv
+	d.KeyImportSymSeed = nil
+}
+
+func (d *keyData_v3) ValidateData(tpm *tpm2.TPMContext, session tpm2.SessionContext) (tpm2.ResourceContext, error) {
```

Comment:
ValidateData for v1 (which is also used by v2) seems to have much better coverage of the error handling than the similar code here

Reply:
I think this was just an oversight - I pulled this code out of the branch I had for the new sealing API, and I hadn't copied all of the tests across yet. I've added some extra tests now.